### PR TITLE
SI-9105 Fix EnclosingMethod for classes defined in lambdas

### PR DESCRIPTION
**Pull request by lrytz**
The reason for an inconsistent EnclosingMethod attribute was the
symbol owner chain. Initially, closure class symbols don't exist, they
are only created in UnCurry (delambdafy:inline). So walking the
originalOwner of a definition does not give yield closure classes.

The commit also fixes uses of isAnonymousClass, isAnonymousFunction
and isDelambdafyFunction in two ways:
1. by phase-travelling to an early phase. after flatten, the name
    includes the name of outer classes, so the properties may become
    accidentally true (they check for a substring in the name)
2. by ensuring that the (destructive) renames during LambdaLift
    don't make the above properties accidentally true. This was in
    fact the cause for SI-8900.

### CODE REVIEW

**Comment by lrytz** [comment]
[`-Ybackend:GenBCode -Ydelambdafy:method` build](https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/1377/) passed, and also a [community build with `-Ybackend:GenASM -Ydelambdafy:inline`](https://jenkins-dbuild.typesafe.com:8499/job/Community-2.11.x-manual/123/).

**Comment by lrytz** [comment]
Review by @adriaanm

File: src/compiler/scala/tools/nsc/transform/LambdaLift.scala [hunk 1]
```diff
@@ -256,15 +256,33 @@ abstract class LambdaLift extends InfoTransform {
           if (originalName.isTypeName) unit.freshTypeName(prefix)
           else unit.freshTermName(prefix)
 
+        /**
+         * Ensures that name mangling does not accidentally make a class respond `true` to any of
+         * isAnonymousClass, isAnonymousFunction, isDelambdafyFunction, e.g. by introducing "$anon".
+         */
+        def ensureNonAnon(prefix: String) = {
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/LambdaLift.scala]: While this seems like a good idea, I worry about doing this in 2.11.x.

File: src/compiler/scala/tools/nsc/transform/LambdaLift.scala [hunk 2]
```diff
@@ -256,15 +256,33 @@ abstract class LambdaLift extends InfoTransform {
...
+          val nestedIn = "$nestedIn"
+          val anoncls = nme.ANON_CLASS_NAME.toString
+          val anonfun = nme.ANON_FUN_NAME.toString
+          val delamda = nme.DELAMBDAFY_LAMBDA_CLASS_NAME.toString
+          prefix.replace(anoncls, nestedIn + anoncls.replace("$", ""))
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/LambdaLift.scala]: I'd introduce constants for `nestedIn + anoncls.replace("$", "")` and friends.

**Comment by retronym** [comment]
Let's add the test case submitted yesterday in [SI-9124](https://issues.scala-lang.org/browse/SI-9124).

**Comment by retronym** [comment]
We should also run the following check over all the JARs we produce to find if we spot any ill-formed class metadata.

``` scala
import scala.reflect.io._

object Test {
  def main(args: Array[String]): Unit = {
    val jarFile = AbstractFile.getDirectory(new java.io.File(args(0)))
    def flatten(f: AbstractFile): Iterator[AbstractFile] =
      if (f.isClassContainer) f.iterator.flatMap(flatten)
      else Iterator(f)

    val classFQNs = flatten(jarFile).filter(_.hasExtension("class")).map(_.path.replaceAllLiterally("/", ".").replaceAll(".class", ""))
    val classes = classFQNs.map(name => Class.forName(name))
    val okay = classes.map(cls => (cls, try {Left(cls.getEnclosingMethod) } catch { case t: Throwable => Right(t) })).filter(_._2.isRight)
    println(okay.mkString("\n"))  
  }
}
```

```
(export V=2.11.4; ~/scala/$V/bin/scalac sandbox/test.scala && ~/scala/$V/bin/scala Test ~/scala/$V/lib/scala-library.jar)
(class scala.reflect.ClassManifestDeprecatedApis$$anonfun$subargs$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.TraversableLike$$anonfun$iterateUntilEmpty$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.TraversableLike$$anonfun$filterImpl$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.immutable.MapProxy$$anon$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.immutable.SetProxy$$anon$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.SeqViewLike$Reversed$$anonfun$createReversedIterator$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.SeqLike$$anon$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.SeqLike$$anonfun$occCounts$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.GenMapLike$$anonfun$liftedTree1$1$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.parallel.ParMapLike$$anon$4,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.parallel.ParMapLike$$anon$3,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.TraversableLike$$anonfun$3,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.TraversableLike$$anonfun$scala$collection$TraversableLike$$sliceInternal$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.mutable.MapProxy$$anon$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.mutable.FlatHashTable$$anonfun$checkConsistent$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.collection.TraversableLike$$anonfun$iterateUntilEmpty$2,Right(java.lang.InternalError: Enclosing method not found))
```

**Comment by lrytz** [comment]
thanks for the test, @retronym. It passes after my patch, but I extended your test a bit and found more issues. I'll look into this. Here's the current version:

``` scala
import scala.reflect.io._

object Test {
  def main(args: Array[String]): Unit = {
    val jarFile = AbstractFile.getDirectory(new java.io.File(args(0)))
    def flatten(f: AbstractFile): Iterator[AbstractFile] =
      if (f.isClassContainer) f.iterator.flatMap(flatten)
      else Iterator(f)

    val classFQNs = flatten(jarFile).filter(_.hasExtension("class")).map(_.path.replace("/", ".").replaceAll(".class$", ""))
    val classes = classFQNs.map(name => Class.forName(name))
    val okay = classes.map(cls => (cls, 
      try { Left(List(cls.getEnclosingMethod, cls.getEnclosingClass, cls.getEnclosingConstructor, cls.getDeclaredClasses)) }
      catch { case t: Throwable => Right(t) })).filter(_._2.isRight)
    println(okay.mkString("\n"))
  }
}
```

It gives a number of messages like this one:

```
(class scala.sys.process.ProcessBuilder$Sink$class,Right(java.lang.IncompatibleClassChangeError: scala.sys.process.ProcessBuilder and scala.sys.process.ProcessBuilder$Sink$class disagree on InnerClasses attribute))
```

**Comment by lrytz** [comment]
almost there.. still have two issues

```
(class scala.tools.nsc.interpreter.JavapClass$fun$u0020with$u0020files$$anonfun$scala$tools$nsc$interpreter$JavapClass$fun$u0020with$u0020files$$lookup$1$1,Right(java.lang.InternalError: Enclosing method not found))
(class scala.tools.nsc.interpreter.JavapClass$ClassLoaderOps$$anonfun$alldirs$1$1,Right(java.lang.InternalError: Enclosing method not found))
```

**Comment by lrytz** [comment]
these are due to value classes, i'll check tomorrow. the rest should be ready for review by @adriaanm (i addressed your feedback from the last review).

**Comment by adriaanm** [comment]
ok!

**Comment by som-snytt** [comment]
Some guys really overdo it with the value classes. They're so good, you can't have just one.

**Comment by retronym** [comment]
I have often wondered if it would have been better to put the value class method bodies in static methods in the value class, rather than in instance methods on the companion. Same story for defaults in constructor. 

Admittedly the status of STATIC methods is a bit grey within the compilation pipeline, but leaving code in its natural habitat rather than shunting it to the companion might avoid a few headaches.

**Comment by lrytz** [comment]
PLS REBUILD/pr-scala@071ea9716e45a7af080c5ea6e85ced38e08d52d6

**Comment by lrytz** [comment]
[-Ybackend:GenBCode -Ydelambdafy:method build](https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/1379/console) OK

**Comment by lrytz** [comment]
fixed the value classes and added @retronym's test

**Comment by retronym** [comment]
We've picked up an inliner warning en route:

```
% diff /localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/presentation/t7678-presentation.log /localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/presentation/t7678.check
@@ -1,2 +1 @@
-warning: there were 5 inliner warnings; re-run with -Yinline-warnings for details
 reload: TypeTag.scala
```

File: src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala [hunk 3]
```diff
@@ -595,7 +595,8 @@ abstract class GenASM extends SubComponent with BytecodeWriters { self =>
           val x = innerClassSymbolFor(s)
           if(x ne NoSymbol) {
             assert(x.isClass, "not an inner-class symbol")
-            val isInner = !x.rawowner.isPackageClass
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala:598]: Is this the only changed file under `src/compiler/scala/tools/nsc/backend/jvm/` that's relevant to the official/default compilation pipeline?

File: src/interactive/scala/tools/nsc/interactive/Global.scala [hunk 4]
```diff
@@ -152,7 +152,7 @@ class Global(settings: Settings, _reporter: Reporter, projectName: String = "")
 
   // don't keep the original owner in presentation compiler runs
   // (the map will grow indefinitely, and the only use case is the backend)
-  override protected def saveOriginalOwner(sym: Symbol) { }
```
Comment (adriaanm) [inline src/interactive/scala/tools/nsc/interactive/Global.scala:155]: can this be problematic for the IDE/Ensime/...?

**Comment by lrytz** [comment]
PLS REBUILD/pr-scala@5ee04deffe699d1d7cd0ec54c05cb692beff2abd

**Comment by lrytz** [comment]
PLS REBUILD/pr-scala@5ee04deffe699d1d7cd0ec54c05cb692beff2abd

**Comment by lrytz** [comment]
PLS REBUILD/pr-scala@9d90035bc67c1a05d3831fd0c15ab032c3637b9f

**Comment by lrytz** [comment]
bye bye kitty

**Comment by lrytz** [comment]
:yum:

**Comment by gkossakowski** [comment]
:grin:

File: src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala (hunk 3 above)
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala:598]: No, the changes in [BCodeAsmCommon](https://github.com/scala/scala/pull/4264/files#diff-1) also apply to GenASM

File: src/interactive/scala/tools/nsc/interactive/Global.scala (hunk 4 above)
Comment (lrytz) [inline src/interactive/scala/tools/nsc/interactive/Global.scala:155]: good point, i'll go back to the original api

**Pull request closed without merging**